### DISCUSSION TITLE
Need help in defing custom language

### DISCUSSION THREAD

**Discussion by kartik-gupta-ij** [microsoft/monaco-editor#3874] [Q&A] [answered]
Hello everyone,

I am new to monaco and I'm having trouble implementing my custom language for a project. I would greatly appreciate some help.

The format of my language is as follows:

```
Method EndPoint
{json Payload (optional)} 
```
I have already created an array of keywords ["POST", "GET", "PUT", "DELETE", "HEAD"] for the HTTP methods. The EndPoint will look something like [collections/startups/points , collections , collections/demo1/points , collections/startups/points].

The {json Payload} part will be optional as per the user's needs.

I am having trouble figuring out how to add custom error checking and auto-completion. I want to be able to check the first word (the HTTP method) and the second word (the endpoint) for unexpected letters and support separate arrays for auto-completion. Additionally, I would like to have a different feature for autocompletion in the JSON body.

Any guidance or advice on how to approach this would be greatly appreciated. Thank you in advance!

current code (for setMonarchTokensProvider  )

```
const Method = ["POST", "GET", "PUT", "DELETE", "HEAD"];

export const Rules = {
    base: 'json',
    Method,
    tokenizer: {
        root: [
            [/@?[a-zA-Z][\w$]*/, {
                cases: {
                    '@Method': 'keyword',

                    '@default': 'variable',
                }
            }],

            [/".*?"/, 'string'],

            [/\/\/.*/, 'comment'],
        ]
    }
}

```

for auto-complete

```

const keywords = ["POST", "GET", "PUT", "DELETE", "HEAD", "collections"];
export const Autocomplete = {
    provideCompletionItems: (model, position) => {
        const suggestions
            =
            [
                ...keywords.map(k => {
                    return {
                        label: "Method:" + k,
                        kind: "monaco.languages.CompletionItemKind.Keyword",
                        insertText: k,
                    };
                })
            ];
        return { suggestions: suggestions };
    }
}

```

**Comment by tamayika** [comment, marked as answer]
You can add error underline by markers.
Please see [example](https://microsoft.github.io/monaco-editor/playground.html?source=v0.37.1#XQAAAAKWBQAAAAAAAABBqQkHQ5NjdMpUO-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWS9TrQEnwCsBo--zFpWcBJnjzBscU8aWOLqQSe9M4rSRp2vK9MmdibnxnF7BoSa8qqj4UDHKDPpFVoi65E1abqBrKTfGHRc16-BJQwB_6G1k7iJcXhF8bGeiFLmR_dUKaFMQsr0YRRd2RmMgwoAmbFlG8lUfDc3XfLVYcWhSkb2iktoFkE-MZrHQnZ3xmaZlHd_U-azcol_9WFj-a_WgYYBBgy3pbYdNoZrv8a3GatD5ETHOiren_STkZag_GgPZhwypLy8MDBWSIxlYd1o04NmjeACMjse8_oaKIgMrGBDTVlJ4y8fym5I90-xXCe-9kupJ6xPeo91SrlEGf9-KREZyXBYm6scxeI8A_5fHEVkkz97RgMt3Gt5QatJUD-RoUa4GDBTLo-xfceW9aT-eKX5hwnkUVz0oUDaKz-rAIwiSZS81jh13JZPEmMl0OV-cN96DXvW2Iy_FOWYSVB8s8m60qExpWPdNxA0aroDwKolT37dXyqDdu1Hpd1QqVJ5uftfVBO1KbQrOHGy30jVVzlnUaiJajGXv4zUqyR510xxjV8h72a5HgOTn4i0VIIdeuTT9U2PC4Xc9U67AafN7LwpTihR1zXF8CUBV0e4uLmwohn-SIiG9ISjKRrUEgGPl0CwRU0PH9jPf6fmSAMgwEEcU-exRzXpBtBzzd9HTX1AcLFxGnSMNuSdhhgV1K-Nrg2LFimNdpJvLZaG-wnO_EJpS8sIjqiCL_znUeag)

Generally, you should create AST parser to provide rich language support.
https://en.wikipedia.org/wiki/Abstract_syntax_tree

**Reply by kartik-gupta-ij** [reply]
@tamayika thanks for the help 
can you please guide me on how to autocomplete for the `Endpoints` and json

Autocomplete should suggest:

next allowed work in API url. Examples

`/collect -> ["collections"]`
`/collections/mycollection/ -> ["points", "snapshots"]`

Allowed fields in JSON:
```
{
  "filter": {
    "|
  }
}
```
should suggest

```

must
should
must_not
```

**Reply by aimhighrana** [reply]
Nice to have thanks
